### PR TITLE
Cache actions that generate symlinks

### PR DESCRIPTION
**Pull request by aalekseyev**
Fix a bug where dune always re-runs all actions that produce symlinks

### CODE REVIEW

File: test/blackbox-tests/test-cases/symlink-targets.t/run.t [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  $ echo a > a
+  $ dune build ./b --display=short
+          bash b
+  $ readlink _build/default/b
+  a
+  $ cat _build/default/b
+  a
+
+# BUG: Dune re-runs the action even though neiter its dependencies, nor its
+# targets were changed.
```
Comment (jeremiedimino) [inline test/blackbox-tests/test-cases/symlink-targets.t/run.t:16]: I suppose this comment should be updated?

**Comment by jeremiedimino** [comment]
We also need a changelog entry BTW

**Comment by aalekseyev** [comment]
Updated the comment and added a changelog entry.

**Review by jeremiedimino — APPROVED** [review]
Looks good!

**Comment by ejgallego** [comment]
Is this related to https://github.com/ocaml/dune/issues/4273 ? I'd bet so.

**Comment by jeremiedimino** [comment]
It would explain the re-execution of the symlinks rules themselves, but not the re-execution of rules that depends on these symlinks. The most likely explanation of #4273 is the know cmt reproducibility issue, which is fixed in a recent version of the compiler.

**Comment by jeremiedimino** [comment]
BTW, this PR seems like a good PR to backport. Adding it to a 2.8.6 milestone.

**Pull request merged**